### PR TITLE
BUG: multivariate_normal returns a pdf for values outside its support when covariance is singular

### LINKED ISSUE #5286: BUG: multivariate_normal returns a pdf for values outside its support when covariance is singular
When a multivariate normal distribution has a singular covariance matrix, its support (i.e. the set of possible values the random variable can take) is restricted to manifold. In this case, the density with respect to the usual Lebesgue measure is not defined. There IS a density with respect to the Hausdorff measure on the support, and this is exactly what stats.multivariate_normal returns. The problem is that it returns a pdf for any input. Specifically, it returns the density of the projection of the input onto the support. I don't think this behaviour is correct. To demonstrate:

```python
>>> import numpy as np
>>> from scipy.stats import multivariate_normal
>>> mean = np.zeros(2)
>>> cov = np.ones((2,2))
>>> rv = multivariate_normal(mean=mean, cov=cov, allow_singular=True)
>>> x = np.array([[0.0, 0.0], [1.0, -1.0], [2.0, -2.0]])
>>> x
array([[ 0.,  0.],
       [ 1., -1.],
       [ 2., -2.]])
>>> rv.pdf(x)
array([ 0.28209479,  0.28209479,  0.28209479])
```

For this distribution, the random variable must lie on the line `y = x`. `[0,0]` is within the support, so that's all well and good. The other two points are not. Technically, the density is simply undefined at these points. For consistency with other scipy.stats distributions, the output should probably be `0.0`.

### PR DESCRIPTION
**Pull request by drpeteb**
Closes #5286.

Before returning a value from `pdf` or `logpdf`, these methods now test that `x` is in the support of the distribution, by removing the projection and seeing if anything is left. The tolerance for this test is the same as that for deciding if the eigenvalues are zero.

The corresponding test, `test_degenerate_distributions` has been modified so that the test points are always in the support. Points guaranteed not to be in the support are also generated and we test that these get a pdf of `0.0`.

### CODE REVIEW

**Comment by drpeteb** [comment]
Dammit. I think this breaks the case when x is multidimensional. Need to replace the `dot`s with `tensordot`s.

**Comment by ev-br** [comment]
@argriffing you are the resident expert, what's your take on this?

**Comment by argriffing** [comment]
> @argriffing you are the resident expert, what's your take on this?

Well @josef-pkt is the expert on the stats infrastructure and @jvkersch is the multivariate normal distribution author.  I think the idea behind this PR makes at least as much sense as the current behavior, but I don't have a strong opinion and I don't think I've ever actually used the singular multivariate normal pdf features.

After reading some of the comments in the issue thread, I kind of agree that the multivariate normal interface is analogous to a hypothetical linalg.inv(M, automatically_use_pseudoinverse_if_nearly_singular=False) which is probably worse than separate inv and pinv, but I see how that discussion would be off-topic for this PR.

**Comment by josef-pkt** [comment]
I'm not an expert on the multivariate implementation. My comments are mostly theoretical or based on my experience in statsmodels. As I commented in another issue, I never went through the details of the multivariate infrastructure and class hierarchy. (I tried to browse a bit right now, but that's not enough.)

The issue #5286 itself seems to have settled on considering the distribution as defined on the (approximate) subspace. 

The changes here look good when reading through the code, but I never actually used the multivariate normal here and don't know about the deprecation issues. (I would consider out of bounds pdf > 0 a bug.) I also don't know much about choosing the actual `eps` that should be used for numerical closeness.

**Comment by drpeteb** [comment]
I'm having second thoughts about this. The tolerances are all messed up.

I added an extra test which checks that random variates sampled from a degenerate distribution using `rvs()` are detected as being in the support by `pdf()` etc. I had to increase the out-of-support tolerance by a factor of 100 to get it to pass. (Which is a horrible hack.)

This means that the numerical precision of the numpy `multivariate_normal` sampling is worse than the tolerance for the eigenvalues when calculating the numerical rank. That was the intuitive method for assessing whether samples are in the support, but I think this means we can't use it. Any thoughts on what we could use instead?

**Comment by argriffing** [comment]
> Any thoughts on what we could use instead?

I'm not sure about tolerances.

Across numpy/scipy there is no standard tolerance to decide numerical rank or even to decide whether a matrix is numerically singular. I put some effort into making sure that these decisions were internally consistent within scipy's `multivariate_normal` so that for example the pseudo-determinant would be computed in a way that uses the same numerical rank as the pseudo-inverse.  But in general I think this is decided on an ad-hoc way basis for most numpy/scipy functions.

**Comment by mdhaber** [comment]
@drpeteb I'd like to help finish you PR gh-5288. Could you select [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)?

**Comment by mdhaber** [comment]
@drpeteb Could you select [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) so I can help finish this up?

**Comment by drpeteb** [comment]
@mdhaber Done. Do you have a plan for how to check whether the point is in the support?

**Comment by mdhaber** [comment]
Thanks! 


> That was the intuitive method for assessing whether samples are in the support, but I think this means we can't use it. 

I think we can still use it. It's been over six years without a perfect solution, so I think something is better than nothing : )

My approach is similar (almost equivalent?) to what's already here, but I was thinking about it from the other direction: project the input vector onto the nullspace, and check whether the magnitude of that projection is within some ad hoc tolerance. This will still return nonzero PDF for some values that probably shouldn't be considered as being in the domain, but I think that's better than returning nonzero values for _all_ inputs that are not in the domain.

**Comment by mdhaber** [comment]
@rkern thoughts on what to do here?

We're trying to resolve the issue demonstrated [here](https://github.com/scipy/scipy/issues/5286#issuecomment-1097518655): when the `cov` matrix of `stats.multivariate_normal` is singular, `pdf` returns the PDF of `x` _projected_ onto the column space of `cov`. The thought is that the PDF should be zero if there is a "substantial" component of `x` in the nullspace of `cov`. 

Personally, I don't think the tolerance for the critical amount of `x` in the nullspace needs to be perfect, but we were hoping to set the threshold such that multivariate random variates produced by `stats.multivariate_normal.rvs`/`np.Generator.multivariate_normal` would always  have nonzero PDF.

I'm a bit surprised by the magnitude of the projection of multivariate random variates onto the nullspace. It is typically ~1e-9.
```python3
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats

rng = np.random.default_rng(934828450)

eig = np.diag([1, 0])
orth = stats.ortho_group(2).rvs(random_state=rng)
V = orth[:, 1]  # spans nullspace
cov = orth @ eig @ orth.T  # random covariance matrix with e-vals 1 and 0
mean = [0, 0]

rvs = rng.multivariate_normal(mean, cov, size=100000, tol=1e-13)
res = rvs @ V
logres = np.log10(res)
plt.hist(logres)
plt.xlabel('log10 nullspace projection magnitude')
plt.ylabel('frequency')
```
image

The tolerance currently in this PR would need to be increased by a factor of ~10  to reliably produce a nonzero PDF for random variates with this `cov`. Thoughts on whether that is OK? (Or would you prefer a substantially different approach to gh-5286?)

**Comment by tirthasheshpatel** [comment]
> This will still return nonzero PDF for some values that probably shouldn't be considered as being in the support, but I think that's better than returning nonzero PDF for _all_ inputs that are not in the support.

I agree. FWIW, bumping the tolerances to work well enough in practice sounds OK.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:172]:
```suggestion
        self.eps = 1e3*eps
```

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel. Let's see if this does the trick for CI.

**Review by tirthasheshpatel — APPROVED** [review]
All tests passing. I can merge tomorrow if there are no more comments.

**Comment by mdhaber** [comment]
This is backward incompatible, and one could argue that this was a feature, not a bug. Let me send an email out first.

**Comment by mdhaber** [comment]
Before this merges, we should actually add a note to the documentation about what the PDF means in the singular case.

**Comment by ilayn** [comment]
I don't know the consequences of the singular covariance but if you want to have a numerically better eps value then it needs to use the `pinvh` eps. Those hardcoded eps numbers should be removed (probably they were tracking the old `pinv2` bounds).

**Comment by ilayn** [comment]
Basically I mean the part starting at line 59 mentioning the old pinvh bounds which we have changed some time ago. But it is apparently used everywhere here and `sparse.linalg` included.

**Comment by mdhaber** [comment]
> Basically I mean the part starting at line 59
So the `_eigvalsh_to_eps` function:
https://github.com/scipy/scipy/blob/2d4167f658f18d8733f525957f11d02b4db7256a/scipy/stats/_multivariate.py#L59

called here:
https://github.com/scipy/scipy/blob/2d4167f658f18d8733f525957f11d02b4db7256a/scipy/stats/_multivariate.py#L159

should be replaced with `val` from:
https://github.com/scipy/scipy/blob/4cf21e753cf937d1c6c2d2a0e372fbc1dbbeea81/scipy/linalg/_basic.py#L1492-L1499

Is that right?

**Comment by ilayn** [comment]
Yes indeed. The key is to get the array size and the dtype affecting the tolerance value instead of fixed numbers.

**Comment by mdhaber** [comment]
Sure. If you're willing to review it, @ilayn, I'll do that in a separate PR, since would affect a lot of other functions. What I'd want to do is just replace the contents of `_eigvalsh_to_eps` with that code from [scipy/scipy/linalg/_basic.py](https://github.com/scipy/scipy/blob/4cf21e753cf937d1c6c2d2a0e372fbc1dbbeea81/scipy/linalg/_basic.py#L1492-L1499). 

Hopefully my addition to the docstring looks OK to define this behavior. I started to define it mathematically in the docstring, but decided to just point to the reference.

@tirthasheshpatel this PR is set if the docstring looks good and tests pass.

**Comment by mdhaber** [comment]
On the other hand @tirthasheshpatel, since we need to fix the squeezing in `multivariate_normal`, and presumably we'll need to create a new distribution for it, should we leave this alone and put the fix into whatever new distribution we create? Or should we just not allow singular covariance for PDF (only RVS)?

**Comment by ilayn** [comment]
> Sure. If you're willing to review it, @ilayn, I'll do that in a separate PR, since would affect a lot of other functions. What I'd want to do is just replace the contents of _eigvalsh_to_eps with that code from scipy/scipy/linalg/_basic.py

No worries, just wanted to mention the eps behavior change. Otherwise no preference on the scheduling or the overhaul context.

**Comment by tirthasheshpatel** [comment]
> On the other hand @tirthasheshpatel, since we need to fix the squeezing in `multivariate_normal`, and presumably we'll need to create a new distribution for it, should we leave this alone and put the fix into whatever new distribution we create? Or should we just not allow singular covariance for PDF (only RVS) in that new distribution?

I think it's better to put this fix in `multivariate_normal` itself and fix the squeeze issue separately. Adding such fixes in a new distribution would be confusing for newer users and the fix would be hidden from users who don't keep an eye on docs/release notes.

**Comment by mdhaber** [comment]
Ok. Feel free to merge if you're happy with it, then.
Another thought, instead of adding all these new distributions, is to raise `FutureWarning`s for in-place changes in behavior. For instance, raise a `FutureWarning` when calling `rvs` that the output shape is going to change in 1.12.0?

**Comment by tirthasheshpatel** [comment]
> Another thought, instead of adding all these new distributions, is to raise `FutureWarning`s for in-place changes in behavior. For instance, raise a `FutureWarning` when calling `rvs` that the output shape is going to change in 1.12.0? Probably a lot less work for us. Downstream libraries need to check version and branch their logic based on that, but it's a lot cleaner than creating all these duplicates (and coming up with good new names)!

This sounds better. Does this mean the actual fix is implemented only after the `FutureWarning` expires?

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:204]:
```suggestion
                assert pdf > 0.0
```

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:209]:
```suggestion
                assert logpdf > -np.inf
```

**Review by tirthasheshpatel — APPROVED** [review]
Had a quick look again and everything LGTM. Two small non-blocking comments. Let me know if you want to commit them or I can merge as-is!

**Comment by mdhaber** [comment]
> Does this mean the actual fix is implemented only after the FutureWarning expires?

Yes. We could prepare a PR before then and merge it only then.

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:204]:
```suggestion
                assert np.all(pdf > 0.0)
```

Oops, X is an array of quantiles.

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:209]:
```suggestion
                assert np.all(logpdf > -np.inf)
```

**Comment by tirthasheshpatel** [comment]
Merged! Thanks @mdhaber @ilayn @drpeteb!

**Pull request merged**
